commit 75a50b92a432fe92e168049a5e7e83a0467df43d
Author: Andy Lutomirski <luto@amacapital.net>
Date:   Thu Oct 10 13:54:07 2013 -0700

    tls: Explicitly handshake connections and optionally notify protocols
    
    SSL_read can transparently negotiate a TLS session, but it only works
    if I/O is actively occuring.  This means that a handshake can complete
    without us noticing.
    
    Rework TLSMemoryBIOProtocol to explicitly handshake connections and add
    ISSLTransport.notifyHandshakeDone to inform interested protocols
    about the handshake status.

diff --git a/twisted/internet/interfaces.py b/twisted/internet/interfaces.py
index 4021e57..1c12c73 100644
--- a/twisted/internet/interfaces.py
+++ b/twisted/internet/interfaces.py
@@ -2152,6 +2152,15 @@ class ISSLTransport(ITCPTransport):
         Return an object with the peer's certificate info.
         """
 
+    def notifyHandshakeDone():
+        """
+        Returns a Deferred that will complete when the initial handshake
+        is done and will errback if the handshake fails.  (Connection
+        loss during the handshake is considered to be a handshake failure.)
+
+        If the handshake is already complete, then the returned Deferred
+        will already be complete.
+        """
 
 class IProcessTransport(ITransport):
     """
diff --git a/twisted/protocols/test/test_tls.py b/twisted/protocols/test/test_tls.py
index 49e3a79..4545713 100644
--- a/twisted/protocols/test/test_tls.py
+++ b/twisted/protocols/test/test_tls.py
@@ -290,11 +290,12 @@ class TLSMemoryBIOTests(TestCase):
         clientFactory = ClientFactory()
         clientFactory.protocol = Protocol
 
-        clientContextFactory, handshakeDeferred = (
+        clientContextFactory, _ = (
             HandshakeCallbackContextFactory.factoryAndDeferred())
         wrapperFactory = TLSMemoryBIOFactory(
             clientContextFactory, True, clientFactory)
         sslClientProtocol = wrapperFactory.buildProtocol(None)
+        handshakeDeferred = sslClientProtocol.notifyHandshakeDone()
 
         serverFactory = ServerFactory()
         serverFactory.protocol = Protocol
@@ -371,6 +372,108 @@ class TLSMemoryBIOTests(TestCase):
                 connectionDeferred])
 
 
+    def test_notifyAfterSuccessfulHandshake(self):
+        """
+        Calling L{TLSMemoryBIOProtocol.notifyHandshakeDone} after a
+        successful handshake should work.
+        """
+        tlsClient, tlsServer, handshakeDeferred, _ = self.handshakeProtocols()
+
+        result = Deferred()
+
+        def check(_):
+            d = tlsClient.notifyHandshakeDone()
+            d.addCallback(result.callback)
+            d.addErrback(result.errback)
+
+        handshakeDeferred.addCallback(check)
+        return result
+
+
+    def test_notifyAfterFailedHandshake(self):
+        """
+        Calling L{TLSMemoryBIOProtocol.notifyHandshakeDone} after a
+        failed handshake should work.
+        """
+        clientConnectionLost = Deferred()
+        clientFactory = ClientFactory()
+        clientFactory.protocol = Protocol
+
+        clientContextFactory = HandshakeCallbackContextFactory()
+        wrapperFactory = TLSMemoryBIOFactory(
+            clientContextFactory, True, clientFactory)
+        sslClientProtocol = wrapperFactory.buildProtocol(None)
+
+        serverConnectionLost = Deferred()
+        serverFactory = ServerFactory()
+        serverFactory.protocol = Protocol
+
+        # This context factory rejects any clients which do not present a
+        # certificate.
+        certificateData = FilePath(certPath).getContent()
+        certificate = PrivateCertificate.loadPEM(certificateData)
+        serverContextFactory = certificate.options(certificate)
+        wrapperFactory = TLSMemoryBIOFactory(
+            serverContextFactory, False, serverFactory)
+        sslServerProtocol = wrapperFactory.buildProtocol(None)
+
+        connectionDeferred = loopbackAsync(sslServerProtocol, sslClientProtocol)
+
+        result = Deferred()
+
+        def fail(_):
+            result.errback(False)
+
+        def check(reason):
+            d = sslClientProtocol.notifyHandshakeDone()
+            if not d.called:
+                result.errback(Exception('notification should be called'))
+                return
+            d.addCallback(fail)
+            d.addErrback(lambda _: result.callback(None))
+
+        sslClientProtocol.notifyHandshakeDone().addCallbacks(fail, check)
+
+        return gatherResults([connectionDeferred, result])
+
+
+    def test_handshakeAfterConnectionLost(self):
+        """
+        Make sure that the correct handshake paths get run after a connection
+        is lost.
+        """
+        clientConnectionLost = Deferred()
+        clientFactory = ClientFactory()
+        clientFactory.protocol = Protocol
+
+        clientContextFactory = HandshakeCallbackContextFactory()
+        wrapperFactory = TLSMemoryBIOFactory(
+            clientContextFactory, True, clientFactory)
+        sslClientProtocol = wrapperFactory.buildProtocol(None)
+
+        serverConnectionLost = Deferred()
+        serverFactory = ServerFactory()
+        serverFactory.protocol = Protocol
+
+        # This context factory rejects any clients which do not present a
+        # certificate.
+        certificateData = FilePath(certPath).getContent()
+        certificate = PrivateCertificate.loadPEM(certificateData)
+        serverContextFactory = certificate.options(certificate)
+        wrapperFactory = TLSMemoryBIOFactory(
+            serverContextFactory, False, serverFactory)
+        sslServerProtocol = wrapperFactory.buildProtocol(None)
+
+        connectionDeferred = loopbackAsync(sslServerProtocol, sslClientProtocol)
+        result = Deferred()
+
+        def checkSide(side):
+            return self.assertFailure(side.notifyHandshakeDone(), Error)
+
+        return gatherResults([connectionDeferred, checkSide(sslClientProtocol),
+                              checkSide(sslServerProtocol)])
+
+
     def test_getPeerCertificate(self):
         """
         L{TLSMemoryBIOProtocol.getPeerCertificate} returns the
diff --git a/twisted/protocols/tls.py b/twisted/protocols/tls.py
index f139c6a..dd17ab3 100644
--- a/twisted/protocols/tls.py
+++ b/twisted/protocols/tls.py
@@ -37,7 +37,7 @@ to run TLS over unusual transports, such as UNIX sockets and stdio.
 
 from __future__ import division, absolute_import
 
-from OpenSSL.SSL import Error, ZeroReturnError, WantReadError
+from OpenSSL.SSL import Error, ZeroReturnError, WantReadError, WantWriteError
 from OpenSSL.SSL import TLSv1_METHOD, Context, Connection
 
 try:
@@ -55,6 +55,7 @@ from twisted.python import log
 from twisted.python._reflectpy3 import safe_str
 from twisted.internet.interfaces import ISystemHandle, ISSLTransport
 from twisted.internet.interfaces import IPushProducer, ILoggingContext
+from twisted.internet import defer
 from twisted.internet.main import CONNECTION_LOST
 from twisted.internet.protocol import Protocol
 from twisted.internet.task import cooperate
@@ -244,14 +245,14 @@ class TLSMemoryBIOProtocol(ProtocolWrapper):
         on, and which has no interest in a new transport.  See #3821.
 
     @ivar _handshakeDone: A flag indicating whether or not the handshake is
-        known to have completed successfully (C{True}) or not (C{False}).  This
-        is used to control error reporting behavior.  If the handshake has not
-        completed, the underlying L{OpenSSL.SSL.Error} will be passed to the
-        application's C{connectionLost} method.  If it has completed, any
-        unexpected L{OpenSSL.SSL.Error} will be turned into a
-        L{ConnectionLost}.  This is weird; however, it is simply an attempt at
-        a faithful re-implementation of the behavior provided by
-        L{twisted.internet.ssl}.
+        complete (C{True}) or not (C{False}).
+
+    @ivar _handshakeError: If the handshake failed, then this will store
+        the reason.  Otherwise it will be C{None}.
+
+    @ivar _handshakeDeferreds: If the handshake is not done, then this
+        is a list of L{twisted.internet.defer.Deferred} instances to
+        be completed when the handshake finishes.
 
     @ivar _reason: If an unexpected L{OpenSSL.SSL.Error} occurs which causes
         the connection to be lost, it is saved here.  If appropriate, this may
@@ -265,6 +266,7 @@ class TLSMemoryBIOProtocol(ProtocolWrapper):
 
     _reason = None
     _handshakeDone = False
+    _handshakeError = None
     _lostTLSConnection = False
     _writeBlockedOnRead = False
     _producer = None
@@ -272,6 +274,7 @@ class TLSMemoryBIOProtocol(ProtocolWrapper):
     def __init__(self, factory, wrappedProtocol, _connectWrapped=True):
         ProtocolWrapper.__init__(self, factory, wrappedProtocol)
         self._connectWrapped = _connectWrapped
+        self._handshakeDeferreds = []
 
 
     def getHandle(self):
@@ -316,15 +319,55 @@ class TLSMemoryBIOProtocol(ProtocolWrapper):
         # Now that we ourselves have a transport (initialized by the
         # ProtocolWrapper.makeConnection call above), kick off the TLS
         # handshake.
-        try:
-            self._tlsConnection.do_handshake()
-        except WantReadError:
-            # This is the expected case - there's no data in the connection's
-            # input buffer yet, so it won't be able to complete the whole
-            # handshake now.  If this is the speak-first side of the
-            # connection, then some bytes will be in the send buffer now; flush
-            # them.
-            self._flushSendBIO()
+        self.__tryHandshake()
+
+
+    def notifyHandshakeDone(self):
+        d = defer.Deferred()
+        if self._handshakeDone:
+            if self._handshakeError is None:
+                d.callback(None)
+            else:
+                d.errback(self._handshakeError)
+        else:
+            self._handshakeDeferreds.append(d)
+        return d
+
+
+    def __tryHandshake(self):
+        """
+        Attempts to handshake.  OpenSSL wants us to keep trying to
+        handshake until either it works or fails (as opposed to needing
+        to do I/O).
+        """
+        while True:
+            try:
+                self._tlsConnection.do_handshake()
+            except WantReadError:
+                self._flushSendBIO()  # do_handshake may have queued up a send
+                return
+            except WantWriteError:
+                self._flushSendBIO()
+                # And try again immediately
+            except Error as e:
+                self._tlsShutdownFinished(Failure())
+                return
+            else:
+                self.__handshakeSucceeded()
+                return
+
+
+    def __handshakeSucceeded(self):
+        """
+        Mark the handshake done and notify everyone.  It's okay to call
+        this more than once.
+        """
+        if not self._handshakeDone:
+            self._handshakeDone = True
+            deferreds = self._handshakeDeferreds
+            self._handshakeDeferreds = None
+            for d in deferreds:
+                d.callback(None)
 
 
     def _flushSendBIO(self):
@@ -349,6 +392,14 @@ class TLSMemoryBIOProtocol(ProtocolWrapper):
         the protocol, as well as handling of the various exceptions which
         can come from trying to get such bytes.
         """
+        # SSL_read can transparently complete a handshake, but we can't
+        # rely on it: if the handshake is done but there's no application
+        # data, then SSL_read won't tell us.
+        if not self._handshakeDone:
+            self.__tryHandshake()
+        if not self._handshakeDone:
+            return  # Save some effort: SSL_read can't possibly work
+
         # Keep trying this until an error indicates we should stop or we
         # close the connection.  Looping is necessary to make sure we
         # process all of the data which was put into the receive BIO, as
@@ -383,14 +434,10 @@ class TLSMemoryBIOProtocol(ProtocolWrapper):
                 self._flushSendBIO()
                 self._tlsShutdownFinished(failure)
             else:
-                # If we got application bytes, the handshake must be done by
-                # now.  Keep track of this to control error reporting later.
-                self._handshakeDone = True
                 ProtocolWrapper.dataReceived(self, bytes)
 
         # The received bytes might have generated a response which needs to be
-        # sent now.  For example, the handshake involves several round-trip
-        # exchanges without ever producing application-bytes.
+        # sent now.  This is most likely to occur during renegotiation.
         self._flushSendBIO()
 
 
@@ -438,6 +485,24 @@ class TLSMemoryBIOProtocol(ProtocolWrapper):
         Called when TLS connection has gone away; tell underlying transport to
         disconnect.
         """
+        if not self._handshakeDone:
+            # This is a handshake failure (either an explicit failure from
+            # OpenSSL or an implicit failure due to a dropped transport
+            # connection).
+            #
+            # Note: Some testcases evilly call _tlsShutdownFinished(None)
+            # before the handshake finishes.  This can't happen in real life
+            # (none of the call sites allow it), so it's okay that we'll
+            # crash if there's actually anyone waiting for notification
+            # of the handshake result.
+            self._handshakeDone = True
+            self._handshakeError = reason
+
+            deferreds = self._handshakeDeferreds
+            self._handshakeDeferreds = None
+            for d in deferreds:
+                d.errback(reason)
+
         self._reason = reason
         self._lostTLSConnection = True
         # Using loseConnection causes the application protocol's
@@ -457,7 +522,7 @@ class TLSMemoryBIOProtocol(ProtocolWrapper):
         """
         if not self._lostTLSConnection:
             # Tell the TLS connection that it's not going to get any more data
-            # and give it a chance to finish reading.
+            # and give it a chance to finish handshaking and/or reading.
             self._tlsConnection.bio_shutdown()
             self._flushReceiveBIO()
             self._lostTLSConnection = True
@@ -532,9 +597,9 @@ class TLSMemoryBIOProtocol(ProtocolWrapper):
                 self._tlsShutdownFinished(Failure())
                 break
             else:
-                # If we sent some bytes, the handshake must be done.  Keep
-                # track of this to control error reporting behavior.
-                self._handshakeDone = True
+                # SSL_write can transparently complete a handshake.  If we
+                # get here, then we're done handshaking.
+                self.__handshakeSucceeded()
                 self._flushSendBIO()
                 alreadySent += sent
 
diff --git a/twisted/web/topfiles/6204.feature b/twisted/web/topfiles/6204.feature
new file mode 100644
index 0000000..1aed481
--- /dev/null
+++ b/twisted/web/topfiles/6204.feature
@@ -0,0 +1 @@
+twisted.internet.interfaces.ISSLTransport now has a notifyHandshakeDone method to request notification when the handshake succeeds or fails.
